flink-core/src/main/java/org/apache/flink/util/JavaGcCleanerWrapper.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Collection;
+
+/**
+ * Java GC Cleaner wrapper.
+ *
+ * <p>A clean operation can be wrapped with the Java GC Cleaner
+ * which will schedule this operation before GC is run for the given owner object (not reachable in user code).
+ * but only if the cleaner has not been already run explicitly by user before.
+ * If the cleaner is run after GC it will not run clean operation again.
+ * This way we guarantee that the clean operation will always run at some point but only once.
+ *
+ * <p>The wrapper looks up the underlying Java GC Cleaner class in different packages
+ */
+public class JavaGcCleanerWrapper {
+	private static final Logger LOG = LoggerFactory.getLogger(JavaGcCleanerWrapper.class);
+
+	private static final String LEGACY_CLEANER_CLASS_NAME = "sun.misc.Cleaner"; // before Java 9
+	private static final String JAVA9_CLEANER_CLASS_NAME = "java.lang.ref.Cleaner"; // since Java 9+
+
+	private static final Collection<CleanerProvider> CLEANER_PROVIDERS =
+		Arrays.asList(LegacyCleanerProvider.INSTANCE, Java9CleanerProvider.INSTANCE);
+	private static final CleanerFactory CLEANER_FACTORY = findGcCleaner();
+
+	private static CleanerFactory findGcCleaner() {
+		Class<?> foundCleanerClass = null;
+		CleanerProvider foundCleanerProviver = null;
+		Throwable t = null;
+		for (CleanerProvider cleanerProvider : CLEANER_PROVIDERS) {
+			try {
+				foundCleanerClass = Class.forName(cleanerProvider.getCleanerClassName());
+				foundCleanerProviver = cleanerProvider;
+				break;
+			} catch (ClassNotFoundException e) {
+				t = ExceptionUtils.firstOrSuppressed(e, t);
+			}
+		}
+
+		if (foundCleanerClass == null) {
+			throw new FlinkRuntimeException(
+				String.format("A proper Java Cleaner class is not found among %s", CLEANER_PROVIDERS),
```

Comment:
```
public static void main(String[] args) {
		System.out.println(Arrays.asList(Foo.INSTANCE));
	}

	private enum Foo {
		INSTANCE
	}
```

output: 
```
[INSTANCE]
```

Reply:
Do you think `CLEANER_PROVIDERS` should be an enum?

Reply:
Error will be `[...] among [INSTANCE, INSTANCE]`